File: source/Assets/Image.cpp
#include "Image.h"
#include "MemoryHelper.h"

using namespace Emission::Assets;

array<unsigned char>^ Image::LoadImageFromPath(String^ path, int* width, int* height)
{
	int comp = 0;
	return Image::LoadImageFromPath(path, width, height, &comp, ColorFormat::Default);
}

array<unsigned char>^ Image::LoadImageFromPath(String^ path, int* width, int* height, ColorFormat colorFormat)
{
	int comp = 0;
	return Image::LoadImageFromPath(path, width, height, &comp, colorFormat);
}

array<unsigned char>^ Image::LoadImageFromPath(String^ path, int* width, int* height, int* comp, ColorFormat colorFormat)
{
	if (path == nullptr) return nullptr;

	stbi_set_flip_vertically_on_load(1);

	unsigned char* ptr = stbi_load(str_to_char_ptr(path), width, height, comp, (int)colorFormat);
	if (ptr == NULL) {
		stbi_image_free(ptr);
		return nullptr;
	}

	int c = (int)colorFormat;
	if (comp == nullptr) {
		if (colorFormat == ColorFormat::Default) {
			stbi_image_free(ptr);
			stbi__err("cannot found image format", "Cannot find image's color format.");
			return nullptr;
		}
	}
	else c = *comp;

	array<unsigned char>^ buffer = gcnew array<unsigned char>(*width * *height * c);
	//pin_ptr<unsigned char> bufptr = &buffer[0];

	//memcpy(bufptr, ptr, buffer->Length);

	System::Runtime::InteropServices::Marshal::Copy((System::IntPtr)ptr, buffer, 0, buffer->Length);

	stbi_image_free(ptr);

	return buffer;
}

array<unsigned char>^ Image::LoadImageFromMemory(array<System::Byte>^ buffer, int* width, int* height)
{
	return Image::LoadImageFromMemory(buffer, width, height, NULL, ColorFormat::Default);
}

array<unsigned char>^ Image::LoadImageFromMemory(array<System::Byte>^ buffer, int* width, int* height, ColorFormat colorFormat)
{
	return Image::LoadImageFromMemory(buffer, width, height, NULL, colorFormat);
}

array<unsigned char>^ Image::LoadImageFromMemory(array<System::Byte>^ stream, int* width, int* height, int* comp, ColorFormat colorFormat)
{
	if (stream->Length == 0) return nullptr;

	pin_ptr<unsigned char> nativeBuffer = &stream[0];
	unsigned char* arr = nativeBuffer;

	stbi_set_flip_vertically_on_load(1);

	unsigned char* ptr = stbi_load_from_memory(arr, stream->Length, width, height, comp, (int)colorFormat);
	if (ptr == NULL) {
		stbi_image_free(ptr);
		return nullptr;
	}

	array<unsigned char>^ buffer = gcnew array<unsigned char>(*width * *height * *comp);
	System::Runtime::InteropServices::Marshal::Copy((System::IntPtr)ptr, stream, 0, stream->Length);

	stbi_image_free(ptr);

	return buffer;
}

File: source/Core/Emission.Core.IO/Image.cs
/*using System.Runtime.InteropServices;
using StbImageSharp;

namespace Emission.Core.IO
{
    public static unsafe class Image
    {
        public static byte[] LoadImageFromPath(string? path, int* width, int* height) => LoadImageFromPath(path, width, height, ColorFormat.Default);
        public static byte[] LoadImageFromPath(string? path, int* width, int* height, ColorFormat colorFormat) => LoadImageFromPath(path, width, height, (int*)0, colorFormat);
        public static byte[] LoadImageFromPath(string? path, int* width, int* height, int* comp, ColorFormat colorFormat)
        {
            if (string.IsNullOrEmpty(path))
                throw new ArgumentNullException(nameof(path));
            
            if (!EFile.Exists(path))
                throw new FileNotFoundException(path);

            return LoadImageFromStream(new StreamReader(path).BaseStream, width, height, comp, colorFormat);
        }
        
        public static byte[] LoadImageFromMemory(MemoryStream? memoryStream, int* width, int* height) => LoadImageFromMemory(memoryStream, width, height, (int*)0, ColorFormat.Default);
        public static byte[] LoadImageFromMemory(MemoryStream? memoryStream, int* width, int* height, int* comp, ColorFormat colorFormat)
        {
            return LoadImageFromStream(memoryStream, width, height, comp, colorFormat);
        }
        
        public static byte[] LoadImageFromStream(Stream? stream, int* width, int* height, int* comp, ColorFormat colorFormat)
        {
            if (stream == null)
                throw new ArgumentNullException(nameof(stream));

            if (!stream.CanRead || stream.Length == 0)
                throw new ArgumentException(nameof(stream));
            
            //stb_image.stb_image.StbiSetFlipVerticallyOnLoad(1);
            StbImage.stbi_set_flip_vertically_on_load(1);

            byte* ptr = StbImage.stbi__load_and_postprocess_8bit(new StbImage.stbi__context(stream), width, height, comp, (int)colorFormat);
            //stb_image.stb_image.StbiLoadFromMemory(MemoryHelper.BytePtrFromByteArray(streamBuffer), streamBuffer.Length, ref w, ref h, ref c, 0); //stb__load_and_postprocess_16bit(&context, width, height, comp, (int)colorFormat);

            if (ptr == (void*)0)
                throw new EmissionException(EmissionException.ERR_IMAGE, "Cannot load image!");

            int w = *width;
            int h = *height;
            int c = (int)colorFormat;
            if (comp == (int*)0)
            {
                if (colorFormat == ColorFormat.Default)
                {
                    Marshal.FreeHGlobal((IntPtr)ptr);
                    throw new EmissionException(EmissionException.ERR_IMAGE,
                        "Cannot load an image because Emission cannot found image's color format!");
                }
            }
            else c = *comp;
            
            byte[] buffer = new byte[w * h * c];
            
            Marshal.Copy((IntPtr)ptr, buffer, 0, buffer.Length);

            Marshal.FreeHGlobal((IntPtr)ptr);
            stream.Dispose();
            
            return buffer;
        }
    }
    
    public enum ColorFormat
    {
        Default = 0x0,
        Grey = 0x1,
        GretAlpha = 0x2,
        RedGreenBlue = 0x3,
        RedGreenBlueAlpha = 0x4
    }
}*/
